R/writeMSData.R

```diff
@@ -0,0 +1,196 @@
+#'
+#' @param verbose `logical(1)` if progress messages should be displayed.
+#'
+#' @param copy `logical(1)` if metadata (data processings, original file names
+#'     etc) should be copied from the original files.
+#' 
+#' @param software_processing optionally provide specific data processing steps.
+#'     See documentation of the `software_processing` parameter of
+#'     [mzR::writeMSData()].
+#'
+#' @author Johannes Rainer
+#'
+#' @md
+#'
+#' @noRd
+.writeMSData <- function(x, files, outformat = c("mzml", "mzxml"),
+                         verbose = isMSnbaseVerbose(),
+                         copy = TRUE,
+                         software_processing = NULL) {
+    ## Check input.
+    outformat <- match.arg(outformat)
+    if (missing(files))
+        stop("'files' is required!")
+    if (length(files) != length(fileNames(x)))
+        stop("length of 'files' has to match the number of samples")
+    if (verbose)
+        message("Writing ", length(files), " ", outformat, " files.")
+    ## Split per file.
+    x_split <- splitByFile(x, f = factor(fileNames(x)))
+    ## Using mapply below - in principle we could then even switch to
+    ## bpmapply to perform parallel saving of files.
+    res <- mapply(FUN = .writeSingleMSData, x_split, files,
+                  MoreArgs = list(outformat = outformat,
+                                  software_processing = software_processing,
+                                  verbose = verbose, copy = copy),
+                  SIMPLIFY = FALSE, USE.NAMES = FALSE)
+}
+
+#' @description `.writeMSData` for a single file.
+#'
+#' @md
+#' 
+#' @noRd
+.writeSingleMSData <- function(msData, file, outformat,
+                               software_processing = NULL,
+                               verbose = TRUE,
+                               copy = is(msData, "OnDiskMSnExp")) {
+    if (verbose)
+        message("Saving file ", basename(file), "...", appendLF = FALSE)
+    file <- path.expand(file)
+    if (file.exists(file))
+        stop("File ", file, " does already exist! Please use a different ",
+             "file name.")
+    ## o get all peak data. Calling spectrapply will apply also all lazy
+    ##   processing steps. We're not calling the `as` method to avoid R
+    ##   having to look through all namespaces to find the function.
+    pks <- spectrapply(msData, FUN = function(sp) cbind(sp@mz, sp@intensity),
+                       BPPARAM = SerialParam())
+        if (is(msData, "OnDiskMSnExp")) {
+            hdr <- fData(msData)
+            ## o Re-calculate stuff we don't have or which might have been
+            ##   changed:
+            new_vals <- do.call(rbind, lapply(pks, function(sp) {
+                cbind(peaksCount = nrow(sp),
+                      totIonCurrent = sum(sp[, 2], na.rm = TRUE),
+                      basePeakMZ = sp[base::which.max(sp[, 2]), 1][1],
+                      basePeakIntensity = max(sp[, 2], na.rm = TRUE))
```

Comment:
Why you calculating the maximum twice? We could simply just call `base::which.max` once and get nearly 20% speed improvement (I guess this is a noncritical function but anyway):
```r
library("msdata")
library("MSnbase")
library("rbenchmark")

m <- readMSData(msdata::proteomics("Top", full.names=TRUE), mode="onDisk")

pks <- spectrapply(m, function(sp)cbind(sp@mz, sp@intensity))

f1 <- function(pks) {
    do.call(rbind, lapply(pks, function(sp) {
        cbind(peaksCount = nrow(sp),
              totIonCurrent = sum(sp[, 2], na.rm = TRUE),
              basePeakMZ = sp[base::which.max(sp[, 2]), 1][1],
              basePeakIntensity = max(sp[, 2], na.rm = TRUE))
}))}

f2 <- function(pks) {
    do.call(rbind, lapply(pks, function(sp) {
        i <- base::which.max(sp[,2])[1]
        cbind(peaksCount = nrow(sp),
              totIonCurrent = sum(sp[, 2], na.rm = TRUE),
              basePeakMZ = sp[i, 1],
              basePeakIntensity = sp[i, 2])
}))}

benchmark(f1(pks), f2(pks), order="relative")
#      test replications elapsed relative user.self sys.self user.child sys.child
# 2 f2(pks)          100   7.791    1.000     7.764    0.004          0         0
# 1 f1(pks)          100   9.447    1.213     9.360    0.008          0         0

all.equal(f1(pks), f2(pks))
# [1] TRUE
```

Reply:
good point.